src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/ILGen.cs

```diff
@@ -574,10 +574,7 @@ internal static void EmitConstant(this ILGenerator il, object value, Type type)
             throw ContractUtils.Unreachable;
         }
 
-        internal static bool ShouldLdtoken(Type t)
-        {
-            return t is TypeBuilder || t.IsGenericParameter || t.GetTypeInfo().IsVisible;
-        }
+        private static bool ShouldLdtoken(Type t) => t.IsGenericParameter || t.GetTypeInfo().IsVisible;
```

Comment:
I've been wondering the same thing. I think the limitation might be with dynamic methods and CompileToMethod would be okay, which would be worth checking if that gets brought to core, especially since CompileToMethod is less flexible about the fallback dynamic methods use when ShouldLdtoken fails.